### ISSUE TITLE
Juggling ASYNC

### ISSUE THREAD

**Issue by Euklidian-Space** [nodeschool/discussions#1673]
- node version: 4.1.1
- npm version:  2.14.4

My code seems to be getting things out of order.  I do not see why.  My code is below.
### Error output

```
Your submission results compared to the expected:

────────────────────────────────────────────────────────────────────────────────

1.  ACTUAL:    "He's got a massive grog mate get a dog up ya whinge. Shazza got us some slabs no dramas he's got a massive bities. Come a spit the dummy bloody gutful of christmas. "
1.  EXPECTED:  "You little ripper his blood's worth bottling no worries he hasn't got a blue. As busy as a pot when gutful of chokkie. It'll be franger to get a dog up ya arvo. She'll be right tucker my stands out like a crook. "

2.  ACTUAL:    "You little ripper his blood's worth bottling no worries he hasn't got a blue. As busy as a pot when gutful of chokkie. It'll be franger to get a dog up ya arvo. She'll be right tucker my stands out like a crook. "
2.  EXPECTED:  "He hasn't got a trackies to as cross as a brumby. It'll be spewin' no dramas grab us a brickie. Shazza got us some freo mate as dry as a brisvegas. "

3.  ACTUAL:    "He hasn't got a trackies to as cross as a brumby. It'll be spewin' no dramas grab us a brickie. Shazza got us some freo mate as dry as a brisvegas. "
3.  EXPECTED:  "He's got a massive grog mate get a dog up ya whinge. Shazza got us some slabs no dramas he's got a massive bities. Come a spit the dummy bloody gutful of christmas. "

4.  ACTUAL:    ""
4.  EXPECTED:  ""


────────────────────────────────────────────────────────────────────────────────

  ✗  Submission results did not match expected!  


     # FAIL  
     Your solution to JUGGLING ASYNC didn't pass. Try again!  
```
### My Code

``` js
 'use strict';

const http = require("http");
let url_list = process.argv.slice(2);

url_list.forEach((url) => {
  http.get(url, (res) => {

    res.setEncoding('utf8');
    let output = '';

    res.on('data', (data) => {
      output += data;
    });

    res.on('end', () => console.log(output));

  }).on('error', (err) => console.log(err.message));

});
```

**Comment by riyadhalnur** [comment]
The thing about the `for` loop is that it doesn't guarantee order when you're iterating over an array. So even though you're getting the right outputs, they're out of order. You will need to keep track of the URLs and make sure they're in the right order before printing them out.

**Comment by Euklidian-Space** [comment]
It is the same case for a normal for(let i = 0; i < length; i++) {...}  loop?  Not just the forEach?

**Comment by riyadhalnur** [comment]
It applies for the `for` loop itself. `.forEach` and `for(...)` is the same, i.e. syntactic sugar.

**Comment by Euklidian-Space** [comment]
While I am trying to figure this out I am coming across another issue.  When I try to push data to an array in the .on('end', ...) callback nothing happens.  The array is still empty.  Here is my code.

'use strict';
const http = require("http");
let url_list = process.argv.slice(2);
let final_data = [];

url_list.forEach((url) => {
  http.get(url, (res) => {

    res.setEncoding('utf8');
    let output = '';

    res.on('data', (data) => {
      output += data;
    });

    res.on('end', () => {
      final_data.push(output);
    });

  }).on('error', (err) => console.log(err.message));

});

console.log(final_data);

**Comment by riyadhalnur** [comment]
This is happening due to async nature of NodeJS. You're waiting on the response to end before you push to the `final_data` array but by that time the program will execute the `console.log(final_data)` because execution is sequential and the program will return to push to the array in the next tick. Try console logging just after the push and you'll see what I mean.

**Comment by Euklidian-Space** [comment]
Aha! I see!  When I console log after the push it actually executes after console.log(final_data).  Thanks @riyadhalnur !

**Comment by Botanic2015** [comment]
Why my code don't work?

```javascript
var http = require('http'),
    final_data = [];

for (var i = 2; i < process.argv.length; ++i) {
  http.get(process.argv[2], (res) => {
  var rezult = "";
  res.on('data', function(data) {
     rezult += data.toString();
  });
  res.on('end', function() {
      final_data[i-2] = rezult;
  });
});
}
final_data.forEach( (data) => console.log(data) );

```
